File: examples/csharp/Cart/src/CartLogic.cs
using Angzarr;
using Examples;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Serilog;

namespace Angzarr.Examples.Cart;

public class CartLogic : ICartLogic
{
    private readonly Serilog.ILogger _logger;

    public CartLogic(Serilog.ILogger logger)
    {
        _logger = logger.ForContext<CartLogic>();
    }

    public CartState RebuildState(EventBook? eventBook)
    {
        var state = CartState.Empty;

        if (eventBook == null || eventBook.Pages.Count == 0)
            return state;

        foreach (var page in eventBook.Pages)
        {
            if (page.Event == null) continue;
            state = ApplyEvent(state, page.Event);
        }

        return state;
    }

    private CartState ApplyEvent(CartState state, Any eventAny)
    {
        var typeUrl = eventAny.TypeUrl;

        if (typeUrl.EndsWith("CartCreated"))
        {
            var evt = eventAny.Unpack<CartCreated>();
            return state with { CustomerId = evt.CustomerId, Status = "active" };
        }

        if (typeUrl.EndsWith("ItemAdded"))
        {
            var evt = eventAny.Unpack<ItemAdded>();
            var newItems = new List<LineItem>(state.Items) { evt.Item };
            return state with { Items = newItems, SubtotalCents = evt.NewSubtotalCents };
        }

        if (typeUrl.EndsWith("QuantityUpdated"))
        {
            var evt = eventAny.Unpack<QuantityUpdated>();
            var updatedItems = state.Items.Select(item =>
                item.ProductId == evt.ProductId
                    ? new LineItem { ProductId = item.ProductId, Name = item.Name, Quantity = evt.NewQuantity, UnitPriceCents = item.UnitPriceCents }
                    : item).ToList();
            return state with { Items = updatedItems, SubtotalCents = evt.NewSubtotalCents };
        }

        if (typeUrl.EndsWith("ItemRemoved"))
        {
            var evt = eventAny.Unpack<ItemRemoved>();
            var remainingItems = state.Items.Where(i => i.ProductId != evt.ProductId).ToList();
            return state with { Items = remainingItems, SubtotalCents = evt.NewSubtotalCents };
        }

        if (typeUrl.EndsWith("CouponApplied"))
        {
            var evt = eventAny.Unpack<CouponApplied>();
            return state with { CouponCode = evt.CouponCode, DiscountCents = evt.DiscountCents };
        }

        if (typeUrl.EndsWith("CartCleared"))
        {
            return state with { Items = new List<LineItem>(), SubtotalCents = 0, CouponCode = "", DiscountCents = 0 };
        }

        if (typeUrl.EndsWith("CartCheckoutRequested"))
        {
            return state with { Status = "checked_out" };
        }

        return state;
    }

    public EventBook HandleCreateCart(CartState state, string customerId)
    {
        if (state.Exists)
            throw CommandValidationException.FailedPrecondition("Cart already exists");

        if (string.IsNullOrWhiteSpace(customerId))
            throw CommandValidationException.InvalidArgument("Customer ID is required");

        _logger.Information("creating_cart {@Data}", new { customer_id = customerId });

        var evt = new CartCreated
        {
            CustomerId = customerId,
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleAddItem(CartState state, string productId, string name, int quantity, int unitPriceCents)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Cart does not exist");

        if (state.IsCheckedOut)
            throw CommandValidationException.FailedPrecondition("Cart already checked out");

        if (quantity <= 0)
            throw CommandValidationException.InvalidArgument("Quantity must be positive");

        if (state.Items.Any(i => i.ProductId == productId))
            throw CommandValidationException.FailedPrecondition("Item already in cart, use UpdateQuantity");

        var item = new LineItem
        {
            ProductId = productId,
            Name = name,
            Quantity = quantity,
            UnitPriceCents = unitPriceCents
        };

        var itemTotal = quantity * unitPriceCents;
        var newSubtotal = state.SubtotalCents + itemTotal;

        _logger.Information("adding_item {@Data}", new { product_id = productId, quantity, new_subtotal = newSubtotal });

        var evt = new ItemAdded
        {
            Item = item,
            NewSubtotalCents = newSubtotal
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleUpdateQuantity(CartState state, string productId, int quantity)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Cart does not exist");

        if (state.IsCheckedOut)
            throw CommandValidationException.FailedPrecondition("Cart already checked out");

        if (quantity <= 0)
            throw CommandValidationException.InvalidArgument("Quantity must be positive");

        var existingItem = state.Items.FirstOrDefault(i => i.ProductId == productId)
            ?? throw CommandValidationException.FailedPrecondition("Item not in cart");

        var oldItemTotal = existingItem.Quantity * existingItem.UnitPriceCents;
        var newItemTotal = quantity * existingItem.UnitPriceCents;
        var newSubtotal = state.SubtotalCents - oldItemTotal + newItemTotal;

        _logger.Information("updating_quantity {@Data}", new { product_id = productId, old_qty = existingItem.Quantity, new_qty = quantity });

        var evt = new QuantityUpdated
        {
            ProductId = productId,
            OldQuantity = existingItem.Quantity,
            NewQuantity = quantity,
            NewSubtotalCents = newSubtotal
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleRemoveItem(CartState state, string productId)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Cart does not exist");

        if (state.IsCheckedOut)
            throw CommandValidationException.FailedPrecondition("Cart already checked out");

        var existingItem = state.Items.FirstOrDefault(i => i.ProductId == productId)
            ?? throw CommandValidationException.FailedPrecondition("Item not in cart");

        var itemTotal = existingItem.Quantity * existingItem.UnitPriceCents;
        var newSubtotal = state.SubtotalCents - itemTotal;

        _logger.Information("removing_item {@Data}", new { product_id = productId, new_subtotal = newSubtotal });

        var evt = new ItemRemoved
        {
            ProductId = productId,
            NewSubtotalCents = newSubtotal
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleApplyCoupon(CartState state, string couponCode, int discountCents)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Cart does not exist");

        if (state.IsCheckedOut)
            throw CommandValidationException.FailedPrecondition("Cart already checked out");

        if (string.IsNullOrWhiteSpace(couponCode))
            throw CommandValidationException.InvalidArgument("Coupon code is required");

        if (!string.IsNullOrEmpty(state.CouponCode))
            throw CommandValidationException.FailedPrecondition("Coupon already applied");

        _logger.Information("applying_coupon {@Data}", new { code = couponCode, discount_cents = discountCents });

        var evt = new CouponApplied
        {
            CouponCode = couponCode,
            DiscountCents = discountCents
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleClearCart(CartState state)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Cart does not exist");

        if (state.IsCheckedOut)
            throw CommandValidationException.FailedPrecondition("Cart already checked out");

        _logger.Information("clearing_cart {@Data}", new { customer_id = state.CustomerId });

        var evt = new CartCleared
        {
            ClearedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleCheckout(CartState state, int loyaltyPointsToUse)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Cart does not exist");

        if (state.IsCheckedOut)
            throw CommandValidationException.FailedPrecondition("Cart already checked out");

        if (!state.Items.Any())
            throw CommandValidationException.FailedPrecondition("Cart is empty");

        var totalCents = state.SubtotalCents - state.DiscountCents;

        _logger.Information("checkout_requested {@Data}",
            new { customer_id = state.CustomerId, total_cents = totalCents, loyalty_points = loyaltyPointsToUse });

        var evt = new CartCheckoutRequested
        {
            CustomerId = state.CustomerId,
            SubtotalCents = state.SubtotalCents,
            DiscountCents = state.DiscountCents,
            LoyaltyPointsToUse = loyaltyPointsToUse,
            RequestedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };
        evt.Items.AddRange(state.Items);

        return CreateEventBook(evt);
    }

    private static EventBook CreateEventBook(IMessage evt)
    {
        var page = new EventPage
        {
            Num = 0,
            Event = Any.Pack(evt),
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        var book = new EventBook();
        book.Pages.Add(page);
        return book;
    }
}

File: examples/cpp/cart/src/cart_logic.cc
#include "cart_logic.hpp"
#include "validation_error.hpp"
#include <algorithm>

namespace cart {

using namespace angzarr;

int32_t CartState::subtotal() const {
    int32_t total = 0;
    for (const auto& item : items) {
        total += item.quantity * item.price_cents;
    }
    return total;
}

int32_t CartState::find_item_index(const std::string& product_id) const {
    for (size_t i = 0; i < items.size(); ++i) {
        if (items[i].product_id == product_id) return static_cast<int32_t>(i);
    }
    return -1;
}

CartState CartLogic::rebuild_state(const angzarr::EventBook* event_book) {
    CartState state;
    if (!event_book || event_book->pages().empty()) return state;

    for (const auto& page : event_book->pages()) {
        if (page.has_event()) {
            state = apply_event(std::move(state), page.event());
        }
    }
    return state;
}

examples::CartCreated CartLogic::handle_create_cart(
    const CartState& state, const std::string& customer_id) {
    if (state.exists()) throw ValidationError::failed_precondition("Cart already exists");
    if (customer_id.empty()) throw ValidationError::invalid_argument("Customer ID is required");

    examples::CartCreated event;
    event.set_customer_id(customer_id);
    return event;
}

examples::ItemAdded CartLogic::handle_add_item(
    const CartState& state, const std::string& product_id,
    int32_t quantity, int32_t price_cents) {
    if (!state.exists()) throw ValidationError::failed_precondition("Cart does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Cart is not active");
    if (product_id.empty()) throw ValidationError::invalid_argument("Product ID is required");
    if (quantity <= 0) throw ValidationError::invalid_argument("Quantity must be positive");
    if (price_cents <= 0) throw ValidationError::invalid_argument("Price must be positive");

    examples::ItemAdded event;
    event.set_product_id(product_id);
    event.set_quantity(quantity);
    event.set_price_cents(price_cents);
    return event;
}

examples::QuantityUpdated CartLogic::handle_update_quantity(
    const CartState& state, const std::string& product_id, int32_t quantity) {
    if (!state.exists()) throw ValidationError::failed_precondition("Cart does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Cart is not active");
    if (state.find_item_index(product_id) < 0) throw ValidationError::failed_precondition("Item not in cart");
    if (quantity <= 0) throw ValidationError::invalid_argument("Quantity must be positive");

    examples::QuantityUpdated event;
    event.set_product_id(product_id);
    event.set_new_quantity(quantity);
    return event;
}

examples::ItemRemoved CartLogic::handle_remove_item(
    const CartState& state, const std::string& product_id) {
    if (!state.exists()) throw ValidationError::failed_precondition("Cart does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Cart is not active");
    if (state.find_item_index(product_id) < 0) throw ValidationError::failed_precondition("Item not in cart");

    examples::ItemRemoved event;
    event.set_product_id(product_id);
    return event;
}

examples::CouponApplied CartLogic::handle_apply_coupon(
    const CartState& state, const std::string& coupon_code, int32_t discount_cents) {
    if (!state.exists()) throw ValidationError::failed_precondition("Cart does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Cart is not active");
    if (coupon_code.empty()) throw ValidationError::invalid_argument("Coupon code is required");
    if (!state.coupon_code.empty()) throw ValidationError::failed_precondition("Coupon already applied");

    examples::CouponApplied event;
    event.set_coupon_code(coupon_code);
    event.set_discount_cents(discount_cents);
    return event;
}

examples::CartCleared CartLogic::handle_clear_cart(const CartState& state) {
    if (!state.exists()) throw ValidationError::failed_precondition("Cart does not exist");
    if (!state.active()) throw ValidationError::failed_precondition("Cart is not active");

    examples::CartCleared event;
    return event;
}

CartState CartLogic::apply_event(CartState state, const google::protobuf::Any& event) {
    const auto& type_url = event.type_url();

    if (type_url.find("CartCreated") != std::string::npos) {
        examples::CartCreated e;
        event.UnpackTo(&e);
        state.customer_id = e.customer_id();
        state.status = CartStatus::Active;
    } else if (type_url.find("ItemAdded") != std::string::npos) {
        examples::ItemAdded e;
        event.UnpackTo(&e);
        int32_t idx = state.find_item_index(e.product_id());
        if (idx >= 0) {
            state.items[idx].quantity += e.quantity();
        } else {
            state.items.push_back({e.product_id(), e.quantity(), e.price_cents()});
        }
    } else if (type_url.find("QuantityUpdated") != std::string::npos) {
        examples::QuantityUpdated e;
        event.UnpackTo(&e);
        int32_t idx = state.find_item_index(e.product_id());
        if (idx >= 0) state.items[idx].quantity = e.new_quantity();
    } else if (type_url.find("ItemRemoved") != std::string::npos) {
        examples::ItemRemoved e;
        event.UnpackTo(&e);
        int32_t idx = state.find_item_index(e.product_id());
        if (idx >= 0) state.items.erase(state.items.begin() + idx);
    } else if (type_url.find("CouponApplied") != std::string::npos) {
        examples::CouponApplied e;
        event.UnpackTo(&e);
        state.coupon_code = e.coupon_code();
        state.discount_cents = e.discount_cents();
    } else if (type_url.find("CartCleared") != std::string::npos) {
        state.items.clear();
        state.coupon_code.clear();
        state.discount_cents = 0;
    } else if (type_url.find("CartCheckoutCompleted") != std::string::npos) {
        state.status = CartStatus::CheckedOut;
    }

    return state;
}

}  // namespace cart
